mlir/lib/Dialect/ArmSME/Transforms/TileAllocation.cpp

```diff
@@ -137,172 +138,612 @@ static ArrayRef<TileMask> getMasks(ArmSMETileType type) {
+    for (OpOperand &operand : terminator->getOpOperands()) {
+      if (isValidSMETileVectorType(operand.get().getType())) {
+        auto copy =
+            rewriter.create<CopyTileOp>(terminator->getLoc(), operand.get());
+        rewriter.modifyOpInPlace(terminator, [&] { operand.assign(copy); });
+      }
+    }
+  }
+}
+
+/// Prepares the IR for tile allocation. It does this by first 'splitting'
+/// conditional branches (see `splitCondBranches`), then inserting tile copies
+/// at branch operations. The conditional branches are split to prevent the
+/// copies needed for them overlapping between the true and false paths of the
+/// branch (see `tile-allocation-copies.mlir` and
+/// `tile-allocation-liveness.mlir` for examples). The copies break up live
+/// ranges and ensure when moving out of SSA the semantics of the program are
+/// persevered.
+void preprocessForTileAllocation(IRRewriter &rewriter,
+                                 FunctionOpInterface function) {
+  splitCondBranches(rewriter, function);
+  insertCopiesAtBranches(rewriter, function);
+}
+
+/// A live range for a (collection of) tile values. A live range is built up of
+/// intervals [start, end) which represent parts of the program where the value
+/// needs to be live (i.e. in an SME virtual tile).
+struct LiveRange {
+  using RangeSet = llvm::IntervalMap<uint64_t, uint8_t, 16,
+                                     llvm::IntervalMapHalfOpenInfo<unsigned>>;
+  using Allocator = RangeSet::Allocator;
+  static constexpr uint8_t kValidLiveRange = 0xff;
+
+  LiveRange(Allocator &allocator)
+      : ranges(std::make_unique<RangeSet>(allocator)) {}
+
+  /// Returns true if this range overlaps with `otherRange`.
+  bool overlaps(LiveRange const &otherRange) const {
+    return llvm::IntervalMapOverlaps<RangeSet, RangeSet>(*ranges,
+                                                         *otherRange.ranges)
+        .valid();
+  }
+
+  /// Unions this live range with `otherRange`, aborts if the ranges overlap.
+  void unionWith(LiveRange const &otherRange) {
+    for (auto it = otherRange.ranges->begin(); it != otherRange.ranges->end();
+         ++it)
+      ranges->insert(it.start(), it.stop(), kValidLiveRange);
+    values.set_union(otherRange.values);
+  }
+
+  /// Inserts an interval [start, end) for `value` into this range.
+  void insert(Value value, unsigned start, unsigned end) {
+    values.insert(value);
+    if (start != end)
+      ranges->insert(start, end, kValidLiveRange);
+  }
+
+  bool empty() const { return ranges->empty(); }
+  unsigned start() const { return ranges->start(); }
+  unsigned end() const { return ranges->stop(); }
+  bool operator<(LiveRange const &other) const {
+    return start() < other.start();
+  }
+
+  ArmSMETileType getTileType() const {
+    return *getSMETileType(cast<VectorType>(values[0].getType()));
+  }
+
+  std::unique_ptr<RangeSet> ranges;
+  SetVector<Value> values;
+  std::optional<unsigned> tileId;
```

Comment:
Could you document these? In particular, what's the relation here? All Values in `values` are live in all ranges of `ranges`? And then one `tileId` is used for Values in `values`?

Related to this, in `coalesceTileLiveRanges` there's this:
```cpp
DenseMap<Value, LiveRange *> liveRanges;
```
That looks like every `Value` is mapped to one `LiveRange`?

Reply:
```
DenseMap<Value, LiveRange *> liveRanges;
```
That's a map from `Value -> LiveRange` (the live range is a pointer as multiple values can point to the same live range, noted below).

Reply:
Added doc comments to the fields now :+1:

Reply:
Thanks, I think I understand where the confusion is coming from. Correct me if I'm wrong ...

One live range contains multiple intervals. Within each interval, only one value from `values` can be live. Now, `LiveRange` (the class) "doesn't care" - it's when it's created we need to make sure that there are no clashes.

Now, there's only one key, `kValidLiveRange`, because this is representing one "live range", right? So, it sounds like it's more like `kLiveRangeID` rather than `kValidLiveRange`? Also, it would be good to add a note that all intervals correspond to one live range, hence one value.

Is this correct?

Reply:
One live range can contain multiple intervals which could be due to:
- Merging the non-overlapping live ranges of multiple values
   - Merging results in live ranges that contain multiple values 
- A value that's live within multiple blocks (i.e. live-ins, each block will have it's own range)

It does not matter how the intervals were created (they just all have to be non-overlapping).  The fact that the intervals are non-overlapping means at any point in a live range only one value is live.

In an `IntervalMap` the keys are intervals. So the mapping is from intervals to `kValidLiveRange` (a dummy value to indicate that interval is part of the live range). We really want a `IntervalSet` (but that does not exist, but this pattern is used elsewhere in MLIR).

Reply:
> We really want a IntervalSet (but that does not exist, but this pattern is used elsewhere in MLIR).

Makes sense. I think that adding this comment and a note that `kValidLiveRange` is just some dummy would be helpful